### PR TITLE
Fix msghistory behavior and tests

### PR DESCRIPTION
**Pull request by Shougo**
https://github.com/vim/vim/pull/16048#issuecomment-2481316507

I have added tests for `msghistory` and fixed bugs.

* when `msghistory` option is changed, the history is truncated

### CODE REVIEW

**Comment by Shougo** [comment]
Hmmm...  It is hard to fix for Windows tests.

File: src/option.c [hunk 1]
```diff
@@ -4924,6 +4925,8 @@ check_num_option_bounds(
 	errmsg = e_invalid_argument;
 	p_mhi = 10000;
     }
+    check_msg_hist();
```
Comment (zeertzjq) [inline src/option.c]: This shouldn't be added here. Add a new `did_set_msghistory()` function and set it in `optiondefs.h`

Comment (Shougo) [reply src/option.c]: Thanks.  I have added it.

**Comment by Shougo** [comment]
Hm.... Can you fix the tests for Windows?

**Comment by Shougo** [comment]
`'Messages maintainer: The Vim Project'` message must be filtered.

File: src/testdir/test_cmdline.vim [hunk 2]
```diff
@@ -4011,4 +4011,36 @@ func Test_cd_bslash_completion_windows()
...
+  call assert_equal(GetMessages(), ['baz', 'foobar'])
+
+  " When the number of messages is 10 and 'msghistory' is changed to 5, is the
+  " number of output lines of :messages 5?
+  set msghistory=10
+  for num in range(1, 10)
+    echomsg num
+  endfor
+  set msghistory=5
+  call assert_equal(len(GetMessages()), 5)
```
Comment (zeertzjq) [inline src/testdir/test_cmdline.vim]:
```suggestion
  call assert_equal(5, len(GetMessages()))
```

File: src/testdir/test_cmdline.vim [hunk 3]
```diff
@@ -4011,4 +4011,36 @@ func Test_cd_bslash_completion_windows()
...
+  endfunction
+
+  " After setting 'msghistory' to 2 and outputting a message 4 times with
+  " :echomsg, is the number of output lines of :messages 2?
+  set msghistory=2
+  echomsg 'foo'
+  echomsg 'bar'
+  echomsg 'baz'
+  echomsg 'foobar'
+  call assert_equal(GetMessages(), ['baz', 'foobar'])
```
Comment (zeertzjq) [inline src/testdir/test_cmdline.vim]:
```suggestion
  call assert_equal(['baz', 'foobar'], GetMessages())
```

File: src/message.c [hunk 4]
```diff
@@ -1039,6 +1035,8 @@ add_msg_hist(
     if (first_msg_hist == NULL)
 	first_msg_hist = last_msg_hist;
     ++msg_hist_len;
+
+    check_msg_hist();
```
Comment (chrisbra) [inline src/message.c:1039]: why was this moved down here?

File: src/option.c [hunk 5]
```diff
@@ -3864,6 +3864,16 @@ did_set_number_relativenumber(optset_T *args UNUSED)
...
 
+/*
+ * Process the updated 'msghistory' option value.
+ */
+    char *
+did_set_msghistory(optset_T *args UNUSED)
+{
+    check_msg_hist();
+    return NULL;
+}
```
Comment (chrisbra) [inline src/option.c:3890]: please move the error checking (max val/min val) here.

File: src/testdir/test_cmdline.vim [hunk 6]
```diff
@@ -4011,4 +4011,36 @@ func Test_cd_bslash_completion_windows()
...
+  set msghistory=10
+  for num in range(1, 10)
+    echomsg num
+  endfor
+  set msghistory=5
+  call assert_equal(len(GetMessages()), 5)
+
+  " Check empty list
+  set msghistory=0
+  call assert_true(empty(GetMessages()))
```
Comment (chrisbra) [inline src/testdir/test_cmdline.vim:4035]: Can you please add this to the option documentation `:h 'msghistory'`
```
Setting it to zero clears the message history.
```

File: src/message.c (hunk 4 above)
Comment (Milly) [reply src/message.c:1039]: If we leave it in its original position, `&msghistory + 1` records will be remembered, right?

Comment (Milly) [reply src/message.c:1039]: In the previous implementation, the constant 200 actually remembered 201 records.

Comment (chrisbra) [reply src/message.c:1039]: ah yes correct

Comment (Shougo) [reply src/message.c:1039]: Yes.  The previous implementation is not correct.

File: src/testdir/test_cmdline.vim (hunk 6 above)
Comment (Shougo) [reply src/testdir/test_cmdline.vim:4035]: Done.

File: src/option.c (hunk 5 above)
Comment (Shougo) [reply src/option.c:3890]: Done.

File: src/testdir/test_cmdline.vim [hunk 7]
```diff
@@ -4011,4 +4011,36 @@ func Test_cd_bslash_completion_windows()
   let &shellslash = save_shellslash
 endfunc
 
+func Test_msghistory()
+  " NOTE: 'Messages maintainer:' message must be filtered
+  function! GetMessages() abort
```
Comment (Milly) [inline src/testdir/test_cmdline.vim]:
```suggestion
  function GetMessages() abort
```

No use `!`, because it would cause an error if already exists.

File: src/testdir/test_cmdline.vim [hunk 8]
```diff
@@ -4011,4 +4011,36 @@ func Test_cd_bslash_completion_windows()
...
+  for num in range(1, 10)
+    echomsg num
+  endfor
+  set msghistory=5
+  call assert_equal(5, len(GetMessages()))
+
+  " Check empty list
+  set msghistory=0
+  call assert_true(empty(GetMessages()))
+
```
Comment (Milly) [inline src/testdir/test_cmdline.vim:4036]:
```suggestion

  delfunction GetMessages
```

File: src/testdir/test_cmdline.vim (hunk 7 above)
Comment (zeertzjq) [reply src/testdir/test_cmdline.vim]: Isn't it the other way around?

Comment (Milly) [reply src/testdir/test_cmdline.vim]: Temporary functions should be cleaned up, if another test forgets to clean up or if there are global helper functions we can catch the overloading error.

Comment (zeertzjq) [reply src/testdir/test_cmdline.vim]: Yes

**Comment by Milly** [comment]
@Shougo `GetMessages` helper is already exists in shred.vim and it's behaviour is same, so we can remove from here.

**Comment by Shougo** [comment]
> @Shougo GetMessages helper is already exists in shred.vim and it's behaviour is same, so we can remove from here.

Oh, I don't know that.  Fixed.

**Comment by Shougo** [comment]
@chrisbra Hi, can you merge the PR before #16068?

**Comment by chrisbra** [comment]
sure, hadn't had time yet. Thanks for pinging me.

**Comment by Shougo** [comment]
Thanks.
